networkx/readwrite/tests/test_text.py

```diff
@@ -0,0 +1,263 @@
+    forest = nx.disjoint_union_all([tree1, tree2])
+    ret = nx.forest_str(forest)
+    print(ret)
+
+    target = dedent(
+        """
+        ╟── 0
+        ╎   ├─╼ 1
+        ╎   │   ├─╼ 3
+        ╎   │   └─╼ 4
+        ╎   └─╼ 2
+        ╎       ├─╼ 5
+        ╎       └─╼ 6
+        ╙── 7
+            ├─╼ 8
+            │   ├─╼ 10
+            │   └─╼ 11
+            └─╼ 9
+                ├─╼ 12
+                └─╼ 13
+        """
+    ).strip()
+    assert ret == target
+
+    tree3 = nx.balanced_tree(r=2, h=2, create_using=nx.DiGraph)
+    forest = nx.disjoint_union_all([tree1, tree2, tree3])
+    ret = nx.forest_str(forest, sources=[0, 14, 7])
+    print(ret)
+
+    target = dedent(
+        """
+        ╟── 0
+        ╎   ├─╼ 1
+        ╎   │   ├─╼ 3
+        ╎   │   └─╼ 4
+        ╎   └─╼ 2
+        ╎       ├─╼ 5
+        ╎       └─╼ 6
+        ╟── 14
+        ╎   ├─╼ 15
+        ╎   │   ├─╼ 17
+        ╎   │   └─╼ 18
+        ╎   └─╼ 16
+        ╎       ├─╼ 19
+        ╎       └─╼ 20
+        ╙── 7
+            ├─╼ 8
+            │   ├─╼ 10
+            │   └─╼ 11
+            └─╼ 9
+                ├─╼ 12
+                └─╼ 13
+        """
+    ).strip()
+    assert ret == target
+
+
+def test_undirected_multi_tree_forest():
+    tree1 = nx.balanced_tree(r=2, h=2, create_using=nx.Graph)
+    tree2 = nx.balanced_tree(r=2, h=2, create_using=nx.Graph)
+    tree2 = nx.relabel_nodes(tree2, {n: n + len(tree1) for n in tree2.nodes})
+    forest = nx.union(tree1, tree2)
+    ret = nx.forest_str(forest, sources=[0, 7])
+    print(ret)
+
+    target = dedent(
+        """
+        ╟── 0
+        ╎   ├── 1
+        ╎   │   ├── 3
+        ╎   │   └── 4
+        ╎   └── 2
+        ╎       ├── 5
+        ╎       └── 6
+        ╙── 7
+            ├── 8
+            │   ├── 10
+            │   └── 11
+            └── 9
+                ├── 12
+                └── 13
+        """
+    ).strip()
+    assert ret == target
+
+
+def test_undirected_tree_str():
+    # Create a directed forest with labels
+    graph = nx.balanced_tree(r=2, h=2, create_using=nx.Graph)
+
+    # arbitrary starting point
+    nx.forest_str(graph)
+
+    node_target0 = dedent(
+        """
+        ╙── 0
+            ├── 1
+            │   ├── 3
+            │   └── 4
+            └── 2
+                ├── 5
+                └── 6
+        """
+    ).strip()
+
+    # defined starting point
+    ret = nx.forest_str(graph, sources=[0])
+    print(ret)
+    assert ret == node_target0
+
+    # defined starting point
+    node_target2 = dedent(
+        """
+        ╙── 2
+            ├── 0
+            │   └── 1
+            │       ├── 3
+            │       └── 4
+            ├── 5
+            └── 6
+        """
+    ).strip()
+    ret = nx.forest_str(graph, sources=[2])
+    print(ret)
+    assert ret == node_target2
+
+
+def test_forest_str_errors():
+    import pytest
+
+    ugraph = nx.generators.complete_graph(3, create_using=nx.Graph)
+
```

Comment:
Should import pytest appear at the module level?
Also, I think we should use `nx.complete_graph` rather than `nx.generators.complete_graph` to reduce maintenance cost of refactoring. Is there an advantage to spelling out the subpackage structure?

Reply:
Predominately local imports is a side effect of my dev style. Local imports don't bother me especially if they are the only function in the module using the imported module. It makes it easier to cut and paste code between modules. That being said, it may be more pythonic to globally import it in this case, and I don't have a strong opinion on it. 

Me using `nx.generators.complete_graph` was a side effect due to how big the networkx global namespace is. I don't always remember what a function is called, but I knew there was a generators module, so it was easier for me to tab-complete to `complete_graph` after I specified the generators module. Its easy enough to drop the generators piece, and I don't think there is a benefit for specifying it as such given the current structure.